### PR TITLE
Remove testcategories from csproj, in relation to the recent changes in buildtools.

### PR DESCRIPTION
**Pull request by Priya91**
Refer dotnet/buildtools@71577daf8e39778fecb5720752fe9e4247a376c8

### CODE REVIEW

**Comment by mellinoe** [comment]
Could you explain what's different about what we are doing now versus what we will do with this change?

**Comment by Priya91** [comment]
@mellinoe : Only the implementation is changing, the behavior remains the same:

1. plain build.cmd will run all tests that are not failing and not outerloop: which is equivalent to, build.cmd /p:WithoutCategories="OuterLoop;failing"
2. To build and run only outerloop, build.cmd /p:WithCategories=OuterLoop
3. To build and run only failing tests, build.cmd /p:WithCategories=failing
4. To build and run any custom category, build.cmd /p:WithCategories=custom or without custom categories, build.cmd /p:WithoutCategories=custom.
5. These categories that are in use in a project, dont have to be specified in TestCategories property in csproj.
6. Xunit will not return error and cause build failure, if it does not discover any tests with the provided category filtering, it just reports a warning.

**Comment by mellinoe** [comment]
I see, so it just keeps us from having to specify the categories in the project file itself, and it figures out the categories from the test files/sources themselves. Sounds good to me.

**Comment by krwq** [comment]
How about tests which are Perf (Category=Perf and OuterLoop)?

**Comment by Priya91** [comment]
@krwq : The perf tests will be marked with OuterLoop attribute, and will be excluded.

**Comment by krwq** [comment]
So there is no way of running only perf tests?

**Comment by Priya91** [comment]
If perf tests are not in OuterLoop, then the Default categories that are not in innerloop can be specified with /p:DefaultNoCategories, and these will be excluded from plain build.cmd. So perf tests need not be marked with OuterLoop, and can be run as build.cmd /p:WithCategories=Perf

**Comment by krwq** [comment]
Perf tests are within OuterLoop. With this change there will be no way of running only Perf tests.
my point is that: OuterLoop=(Perf + LongRunningTests + whatever else)
if you say WithCategories=Perf you will get:
-trait category=Perf -notrait=OuterLoop;failing which is not gonna run anything.

Also to add perf tests or other category of tests now you need to update build script and you did not have to previously

**Comment by Priya91** [comment]
@krwq : No, we have agreed to separate perf and outerloop. Tests that are marked with perf, will not be marked with outerloop. Perf is a separate category, that is different from outerloop. so the structure i have in mind is:
A test can be Outerloop with,
1. category 1
2. category 2
3. If cannot have a specific category name, mark it with outerloop.

Does this work for you, or what are the scenarios that's missing with this structure?

**Comment by krwq** [comment]
I got no opinion on this. I believe we just went back to the solution we had on the day 1.

File: dir.props [hunk 1]
```diff
@@ -2,8 +2,8 @@
   <Import Condition="Exists('..\dir.props')" Project="..\dir.props" />
   <!-- Build Tools Version -->
   <PropertyGroup>
-    <BuildToolsVersion>1.0.25-prerelease-00033</BuildToolsVersion>
-    <DnxVersion>1.0.0-beta5-11562</DnxVersion>
+    <BuildToolsVersion>1.0.25-prerelease-00037</BuildToolsVersion>
+    <DnxVersion>1.0.0-beta5-11580</DnxVersion>
```
Comment (weshaggard) [inline dir.props]: Why did you also need to update the version of dnx?

File: src/System.Xml.XDocument/tests/events/System.Xml.XDocument.Events.Tests.csproj [hunk 2]
```diff
@@ -9,7 +9,6 @@
     <AssemblyName>System.Xml.XDocument.Events.Tests</AssemblyName>
     <SignAssembly>false</SignAssembly>
     <ProjectGuid>{C560E194-5B14-4112-ABC6-3208491E53E6}</ProjectGuid>
-    <TestCategories>OuterLoop</TestCategories>
```
Comment (weshaggard) [inline src/System.Xml.XDocument/tests/events/System.Xml.XDocument.Events.Tests.csproj]: We should continue supporting marking tests projects as OuterLoop and then completely skip running them when we are only doing InnerLoop test runs.

**Comment by weshaggard** [comment]
I actually agree with @krwq we shouldn't be moving back to a place were we always run the tests for every test project even if there isn't any tests to run for that project.

File: dir.props (hunk 1 above)
Comment (Priya91) [inline dir.props]: Resolving merge conflict.. I must have pushed the wrong branch. Will fix this.

**Comment by eatdrinksleepcode** [comment]
@WesHaggard Is the goal there to avoid the overhead of running xunit?

File: src/System.Xml.XDocument/tests/events/System.Xml.XDocument.Events.Tests.csproj (hunk 2 above)
Comment (krwq) [inline src/System.Xml.XDocument/tests/events/System.Xml.XDocument.Events.Tests.csproj]: I think it is too late for that: https://github.com/dotnet/buildtools/pull/144
This pretty much reverted all my changes.

**Comment by krwq** [comment]
I can see that adding TestCategories to every single project is annoying but we can make this optional by setting default value of TestCategories to all known categories: if someone doesn't set it it will be ok but if he does it will skip the project if there are no tests to run. Additionally we can make a tool which updates it automatically.

**Comment by Priya91** [comment]
From the conversation above, and the current state of the script, this is the summary:

Test build properties

Xunit related:

1. **WithCategories**: Used to specify the tests that have these traits to be run
2. **WithoutCategories**: Used to specify the tests that do not have these traits to be run
3. **DefaultNoCategories**: Values hardcoded in build script: {OuterLoop;failing}, When more categories like perf, stress, etc are supported pass these categories (i.e, traits not be run in daily dev runs) to this property. 
4. **TestCategories - Proposed**: Set at project level disable entire project based on category. Check if !(TestCategories has WithCategories) Or (WithCategories=='' And (TestCategories hasNot InnerLoop Or TestCategories=="") is true  then project has no tests to run, so don't call Test target. The problem that comes with this is, all test projects should now be added the TestCategories property, least value to be InnerLoop, note Innerloop is just a logical entity, not an actual xunit trait. This is not easily discoverable, and I wouldn't want it to be the dev responsibility to know this. If this is what is preferred, then we need to automate this, which is going to add more overhead to build.
To overcome this, I propose either:
  - Changing the TestCategories to **RunTestsForCategories** or change **RunTestsForProject in {category; true; false}**, which will invoke test target only if  (WithCategories == "" And WithoutCategories != RunTestsForCategories) Or (WithCategories != '' And RunTestsForCategories == '') Or (RunTestsForCategories != '' And RunTestsForCategories has WithCategories), this makes it possible to not have this property over every project, or deal with "InnerLoop" shadow concept.In this property you only specify those actual xunit categories from attributes, (Innerloop not included), for which you want to run tests for this project.
  - Why is running xunit not preferred when there are no tests to be run? This seems ok, considering the other option of having another project level xunit specific build property. I don't see any hit in performance, and no build errors.

Others:

1. **UnsupportedPlatforms** - Any value from {Windows_NT;OSX;Linux} set at project level to disable this test project on the specified platforms.
2. **RunTestsForProject** - Usually set at the project level to disable the tests for a single project.
3. **CoverageEnabledForProject** - Usually set at the project level to disable code coverage for a single project. 
4. **SkipTests** - Usually set at the root level for builds that want to disable all the tests.

What are your thoughts? @weshaggard @krwq

**Comment by weshaggard** [comment]
>Why is running xunit not preferred when there are no tests to be run? This seems ok, considering the other option of having another project level xunit specific build property. I don't see any hit in performance, and no build errors.

We have a lot of test projects mostly InnerLoop right now and the number will just continue to grow as we fill in more things like Perf/OuterLoop tests so we need to be able to correctly filter out the projects and only do the work that is necessary. While I agree right now it doesn't appear to take much time to run xunit on a project with no tests to run, it will however add up across the larger repo and we should try to avoid it. I'm also thinking ahead because there will come a time were RunTestForProject will be doing more than just the one xunit run as we enable testing for other things like full framework and .NET Native. In particular when we enable .NET Native it will really put a hit on the amount of time to do the testing process and we will want to avoid that hit if the test project isn't applicable. 

Things like these are the sort of requirements I think are worth capturing in a document\wiki so we understand why the system is the way it is and we can ensure that it is meeting all of our requirements.

**Comment by krwq** [comment]
I just tried to run all tests using /p:WithCategories="InnerLoop;OuterLoop" but it seems to run ONLY OuterLoop now

**Comment by Priya91** [comment]
@krwq :That's the expected behaviour and always has been so, this corresponds to xunit.console.netcore.exe -trait category=OuterLoop, i dont understand what you mean?

**Comment by Priya91** [comment]
I would like to reach to a conclusion on this, as this has been in discussion for a long time now. I will add RunTestsForCategories (instead of TestCategories, and no dependency on InnerLoop) that will be added on projects, to specify for what categories this project should be enabled. As @weshaggard points out, since there is more functionality to be added to RunTestForProject, i'll leave it unmodified. Does this sound good?

**Comment by eatdrinksleepcode** [comment]
It feels like we are trying to prematurely optimize the build. We are concerned that calling xunit unnecessarily will add too much overhead to the build in the future, but we don't actually have any evidence of that at this point. I would be in favor of keeping the build simple until we know that it needs to be more complicated.

**Comment by weshaggard** [comment]
@eatdrinksleepcode I agree that may be what it appears but I can tell you that this is just a step in the direction of what we are going to have to do in order to complete our move from our internal infrastructure to the open source infrastructure. Most of my guidance here is based on what we are doing internally and knowing what is coming. 

@Priya91 given you have already made the changes to the buildtools I don't have any issues with merging this (except you should remove your buildtools and dnx version bump as they aren't needed). However I want you to file an issue to start a discussion about how we want our test system to work and to gather a list of requirements/scenarios we need. Once we have a good idea of that then we can publish it on the wiki so that we can have an agreement on what it is we are building and we won't need to have the continued discussions on the various PR's/Issues.

**Comment by mellinoe** [comment]
> However I want you to file an issue to start a discussion about how we want our test system to work and to gather a list of requirements/scenarios we need

I definitely second this. To be honest, I still don't know what the expected behavior of tests is supposed to be, and I'm still not clear on how different traits/categories etc. behave, and what our plan for them long-term is. It would really help to plan this out and write everything down.

**Comment by krwq** [comment]
```
That's correct, and this has been the behavior always
```

That was actually working before your changes to buildtools

**Comment by eatdrinksleepcode** [comment]
@weshaggard To be honest, it is difficult to form a response to "I know something you don't know." I recognize that you are not deliberately withholding anything, but given the information I have right now, I stand by my statement that trying to determine whether there are any tests to run without executing the test runner is complicating the build for no apparent benefit. As things change, if executing the test runner when there are no tests to run is found to be a significant source of overhead, then we can look into reducing that overhead. Not executing the test runner is one potential way to do that, but it is not the only way. What if we could execute the test runner in-process with the build? That would speed up testing across the board and possibly make the overhead of executing the test runner small enough that there is no point in avoiding it. But we can't reasonably make that determination now when we have no data to base a decision on.

> > However I want you to file an issue to start a discussion about how we want our test system to work and to gather a list of requirements/scenarios we need

> I definitely second this.

I'll third this. It is related to dotnet/issues@125: we need tests for the BuildTools that show how it is intended to be used (I have made some progress on that, I just need to find some time to put the PR together).

**Comment by Priya91** [comment]
Merging this PR, and opening an issue to discuss further.

**Comment by Priya91** [comment]
Merging this PR, and opening an issue for further discussion.

**Comment by weshaggard** [comment]
@eatdrinksleepcode I can definitely understand your point of view and I do appreciate you keeping us honest and it is certainly not my intent to keep you or the community in the dark. We are simply trying to build this system up as we go but we do have a fair amount of prior art we are trying to leverage where it makes sense. I also realize there are potentially other possible solutions to solving the issue and perhaps I'm being a little too prescriptive on the solution without fully clarifying the requirements. Hopefully we can use issue #1477 to help nail down the requirements and scenarios and we can come up with the right solution. 

As always we appreciate your feedback and contributions so keep them coming @eatdrinksleepcode.

**Pull request merged**